### PR TITLE
Feature/mvp feed screen

### PR DESCRIPTION
**Pull request by Saik0s**
**This PR is the first step in the creation of a feed screen**

Here is a short overview of changes:
- Simplified project configurations
- New project structure with extracted Data, Domain, and Utilities
- Add some basic components like a splash screen, main flow controller, etc
- Feed screen which retrieves posts from r/popular and displays titles of these posts

### CODE REVIEW

File: Scripts/cpd [hunk 1]
```diff
@@ -10,8 +10,12 @@ cd "${0%/*}/.."
 pmd cpd \
 	--files Projects/Application/Sources \
 	--exclude Projects/Application/Sources/Generated \
-	--minimum-tokens 50 --language swift --encoding UTF-8 \
-	--format net.sourceforge.pmd.cpd.XMLRenderer --failOnViolation true >cpd-output.xml
+	--minimum-tokens 50 \
+    --language swift \
+    --encoding UTF-8 \
+	--format net.sourceforge.pmd.cpd.XMLRenderer \
+    --failOnViolation false \
```
Comment (rencevio) [inline Scripts/cpd:17]: Indentation is a bit messed up here.

Comment (rencevio) [inline Scripts/cpd:17]: Could you explain why `failOnViolation` is false? Isn't that what we want?

File: Projects/Domain/Sources/Services/SubredditService.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    public init(dataSource: SubredditDataSource) {
+        self.dataSource = dataSource
+    }
+
+    public func getPosts(for subredditName: String) -> Observable<[Post]> {
+        return dataSource
+            .fetchPosts(for: subredditName)
+            .asObservable()
+            .map { $0.data.children?.map(DefaultSubredditService.convertPostContainerToDomainPost) }
+            .map { $0 ?? [] }
```
Comment (rencevio) [inline Projects/Domain/Sources/Services/SubredditService.swift:30]: Not sure I like that way of dealing with optionality. Why not filter them out instead (e.g. with `flatMap`; preferably using something like RxOptional or writing our own extension)?

File: Projects/Data/Sources/Entities/Resolutions.swift [hunk 3]
```diff
@@ -0,0 +1,12 @@
+//
+// Created by Igor Tarasenko on 2019-02-07.
+// Licensed under the MIT license
+//
+
+import Foundation
+
+public struct Resolution: Codable {
```
Comment (rencevio) [inline Projects/Data/Sources/Entities/Resolutions.swift]: The file is named `Resolutions`, yet the only type it defines is `Resolution`. Typo or are we planning to define more types here?

File: Projects/Data/Sources/Entities/Post.swift [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        permalink = try values.decodeIfPresent(String.self, forKey: .permalink)
+        parentWhitelistStatus = try values.decodeIfPresent(String.self, forKey: .parentWhitelistStatus)
+        stickied = try values.decodeIfPresent(Bool.self, forKey: .stickied)
+        url = try values.decodeIfPresent(String.self, forKey: .url)
+        subredditSubscribers = try values.decodeIfPresent(Int.self, forKey: .subredditSubscribers)
+        createdUTC = try values.decodeIfPresent(Int.self, forKey: .createdUTC)
+        media = try values.decodeIfPresent(Media.self, forKey: .media)
+        isVideo = try values.decodeIfPresent(Bool.self, forKey: .isVideo)
+    }
+    // swiftlint:enable identifier_name
```
Comment (rencevio) [inline Projects/Data/Sources/Entities/Post.swift]: Why is this here? It is enabled by default anyway.

File: Projects/Data/Sources/Entities/Post.swift [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        case authorPatreonFlair = "author_patreon_flair"
+        case authorFlairTextColor = "author_flair_text_color"
+        case permalink = "permalink"
+        case parentWhitelistStatus = "parent_whitelistStatus"
+        case stickied = "stickied"
+        case url = "url"
+        case subredditSubscribers = "subreddit_subscribers"
+        case createdUTC = "created_utc"
+        case media = "media"
+        case isVideo = "is_video"
```
Comment (rencevio) [inline Projects/Data/Sources/Entities/Post.swift:203]: Great job implementing the models! 👍 
However, could you clarify where did you take those coding keys from? Was it retrieved from the `get` request that you did or from some documentation piece?
I'm asking because I wasn't able to find that list on official API documentation or by googling.

File: Projects/Application/Sources/Common/Extensions/UIViewController+Child.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import UIKit
+
+extension UIViewController {
+    func addAndMoveChild(_ viewController: UIViewController) {
+        addChild(viewController)
+        view.addSubview(viewController.view)
+        viewController.didMove(toParent: self)
+    }
+
+    func removeChild(_ viewController: UIViewController) {
```
Comment (rencevio) [inline Projects/Application/Sources/Common/Extensions/UIViewController+Child.swift]: This function seems weird to me. I don't see any usage of the instance on which it was called, but only the instance (`viewController`) that was passed.
Perhaps adding some check like `viewController.parent == self` will clarify things?

File: Projects/Application/Sources/Flow/AppFlowController.swift [hunk 7]
```diff
@@ -3,26 +3,57 @@
...
-        super.init(nibName: nil, bundle: nil)
+        let reactor = SplashViewReactor()
+        reactor.callback
+            .bind { [weak self] callback in
+                switch callback {
+                case .didFinishLoading:
+                    self?.showMainFlow()
+                }
+            }
+            .disposed(by: disposeBag)
```
Comment (rencevio) [inline Projects/Application/Sources/Flow/AppFlowController.swift:32]: Looks awesome! 🥇

**Review by rencevio — CHANGES_REQUESTED** [review]
Amazing job! Apart from comments and remarks, it is ready for merging :)

File: Scripts/cpd (hunk 1 above)
Comment (Saik0s) [reply Scripts/cpd:17]: It is still going to give us warnings that need to be fixed. If it does not fail, we can continue script execution so PHP script can parse output and print warnings.

File: Projects/Data/Sources/Entities/Post.swift (hunk 5 above)
Comment (Saik0s) [reply Projects/Data/Sources/Entities/Post.swift:203]: Just reverse engineering, could not find all of the info public. According to some Reddit posts, not all of the fields are documented.

File: Projects/Domain/Sources/Services/SubredditService.swift (hunk 2 above)
Comment (Saik0s) [reply Projects/Domain/Sources/Services/SubredditService.swift:30]: I still want to receive `onNext` event, even if there are no results. Don't focus on service right now; this is just a dumb implementation to have something to work with.

File: Scripts/cpd (hunk 1 above)
Comment (Saik0s) [reply Scripts/cpd:17]: Fixed

File: Projects/Data/Sources/Entities/Resolutions.swift (hunk 3 above)
Comment (Saik0s) [reply Projects/Data/Sources/Entities/Resolutions.swift]: Fixed

File: Projects/Data/Sources/Entities/Post.swift (hunk 4 above)
Comment (Saik0s) [reply Projects/Data/Sources/Entities/Post.swift]: Fixed

File: Projects/Application/Sources/Common/Extensions/UIViewController+Child.swift (hunk 6 above)
Comment (Saik0s) [reply Projects/Application/Sources/Common/Extensions/UIViewController+Child.swift]: Fixed

**Review by rencevio — APPROVED** [review]
